README
NOTE
====

In all of the example code here and in the API you need to namespace the object like `GCal4Ruby::Calendar` instead of just `Calendar`

=GCal4Ruby
==Introduction
GCal4Ruby is a full featured wrapper for the google calendar API.  GCal4Ruby implements
all of the functionality available through the Google Calnedar API, including permissions,
attendees, reminders and event recurrence.  

==Author and Contact Information
This branch was made by Matt Gornick (http://github.com/mgornick) to implement Google AuthSub into the original 
GCal4Ruby that was created and is maintained by {Mike Reich}[mailto:mike@seabourneconsulting.com] 
and is licenses under the GPL v2.  Feel free to use and update, but be sure to contribute your
code back to the project and attribute as required by the license.
===Website
http://rubyforge.org/projects/gcal4ruby/

===Example Application
http://github.com/mgornick/studentplanr

==Description
GCal4Ruby has three major components: the service, calendar and event objects.  Each service
has many calendars, which in turn have many events.  Each service is the representation of a
google account, and thus must be successfully authenticated using valid Google Calendar
account credentials.  

==Examples
Below are some common usage examples.  For more examples, check the documentation.
===Service
1. Authenticate
    service = Service.new
    service.authenticate("user@gmail.com", "password")

2. Authenticate with Google AuthSub
		# your authentication controller (example is index action in the home controller):
			scope = 'http://www.google.com/calendar/feeds/'
	    next_url = 'http://127.0.0.1:3000/home/authenticated'
	    secure = false  # set secure = true for signed AuthSub requests
	    session = true
	    @authsub_link = GData::Auth::AuthSub.get_url(next_url, scope, secure, session)

		# in your next_url controller as defined by your AuthSub link (in example would be the authenticated action)
			@token = params[:token]
  		client = GData::Client::Calendar.new
  		client.authsub_token = params[:token] # extract the single-use token from the URL query params
  		session[:token] = client.auth_handler.upgrade()
  		client.authsub_token = session[:token] if session[:token]
  		
  		service= GCal4Ruby::Service.new
  		service.authsub_authenticate(session[:token], "user@gmail.com")

3. Get Calendar List
    calendars = service.calendars

===Calendar
All usages assume a successfully authenticated Service.
1. Create a new Calendar
    cal = Calendar.new(service)

2. Find an existing Calendar
    cal = Calendar.find(service, "New Calendar", :first)

3. Find all calendars containing the search term
    cal = Calendar.find(service, "Soccer Team")

4. Find a calendar by ID
    cal = Calendar.find(service, id, :first)
===Event
All usages assume a successfully authenticated Service and valid Calendar.
1. Create a new Event
    event = Event.new(calendar)
    event.title = "Soccer Game"
    event.start = Time.parse("12-06-2009 at 12:30 PM")
    event.end = Time.parse("12-06-2009 at 1:30 PM")
    event.where = "Merry Playfields"
    event.save

2. Find an existing Event
    event = Event.find(cal, "Soccer Game", {:scope => :first})

3. Find all events containing the search term
    event = Event.find(cal, "Soccer Game")

4. Create a recurring event for every saturday
    event = Event.new(calendar)
    event.title = "Baseball Game"
    event.where = "Municipal Stadium"
    event.recurrence = Recurrence.new
    event.recurrence.start = Time.parse("13-06-2009 at 4:30 PM")
    event.recurrence.end = Time.parse("13-06-2009 at 6:30 PM")
    event.recurrence.frequency = {"weekly" => ["SA"]}
    event.save 

5. Create an event with a 15 minute email reminder
    event = Event.new(calendar)
    event.title = "Dinner with Kate"
    event.start = Time.parse("20-06-2009 at 5 pm")
    event.end = Time.parse("20-06-209 at 8 pm")
    event.where = "Luigi's"
    event.reminder = {:minutes => 15, :method => 'email'}
    event.save

6. Create an event with attendees
    event = Event.new(calendar)
    event.title = "Dinner with Kate"
    event.start = Time.parse("20-06-2009 at 5 pm")
    event.end = Time.parse("20-06-209 at 8 pm")
    event.attendees => {:name => "Kate", :email => "kate@gmail.com"}
    event.save


CHANGELOG
#=CHANGELOG
#==version 0.3.1_OAuth
#* Git repo by tvongaza to implement OAuth authentication into the GCal4Ruby gem
#==version 0.3.1_AuthSub
#* Git repo by mgornick to implement Authsub authentication into the GCal4Ruby gem
#==version 0.3.1
#* Fixed Event.find method to work with secondary calendars and google apps accounts.  Fix provided by groesser3.
#* Added max results to Calendar.find.
#==version 0.3.0
#* Rewrote Event.find to improve performance significantly.
#* Added improvements to event recurrence handling, including loading existing recurrences, changing recurring events to non recurring and vice versa.
#* Added support for initialization attributes to Event, Calendar, Service and Recurrence.
#* Fixed query string typo in Event.find.  Fix provided by nat.lownes.
#==version 0.2.11
#* Added support for GML elements in calendar events.  Fix provided by nat.lownes.
#* Fixed event status bug where cancelled events were marked confirmed.  Fix provided by rifraf.
#==version 0.2.10
#* Fixed library support for Google Hosted Apps accounts running in forced SSL mode.
#==version 0.2.9
#* Fixed small SSL redirect bug due to variable misnaming in base.rb.  Fix provided by JohnMetta
#==version 0.2.8
#* Merged changes from edruder and h13ronim from the unofficial github repo - http://github.com/h13ronim/gcal4ruby/commits/master
#==version 0.2.7
#* Added fix for finding events in calendars that have more than 25 entries
#==version 0.2.6
#* Added fix for updated google calendar XML per http://cookingandcoding.wordpress.com/2009/06/08/new-ruby-google-calendar-api-gem-gcal4ruby/#comment-183
#==version 0.2.5
#* Added calendar color support to to_iframe methods in Calendar and Service.
#==version 0.2.4
#* Fixed bug with ACL check in Calendar#load
#==version 0.2.3
#* Implemented to_iframe method for calendars and services to output embeddable iframe text.
#* Added switch to turn off ACL check for public calendars.  Can increase effeciency if turned off.
#==version 0.2.2
#* Fixed URL encoding problem in Event.find method.
#* cleaned up Event.find method to allow for finding events by id
#* updated Calendar.find method to add params hash
#* added 'published', 'updated', and 'edited' attributes
#==version 0.2.1
#* fixed Event.find calendar specification
#==version 0.2.0
#* Fixed redirect URI query params problem
#* Updated syntax for finding events to include most google api parameters,  Non-backwards compatible.
#==version 0.1.4
#* Added additional search criteria for Event.find
#==version 0.1.3
#* Added support for authenticating with Google Hosted Apps accounts
#* Added flag to indicate whether a calendar is editable
#* Added handling to gracefully throw error when trying to create event on a non-editable (shared) calendar
#==version 0.1.2
#* Fixed to_xml dump problem with hidden and selected attributes
#==version 0.1.1
#* Added all_day indicator to event to indicate an all day event
#==version 0.1.0
#* Initial Version

lib/gcal4ruby/base.rb
require "rexml/document"
require "cgi"
require "uri"
require "net/http"
require "net/https"
require "open-uri"
require "nkf"
require "time"

Net::HTTP.version_1_2

# GCal4Ruby is a full featured wrapper for the google calendar API

# =Usage:

module GCal4Ruby
  
  CALENDAR_XML = "<entry xmlns='http://www.w3.org/2005/Atom' 
       xmlns:gd='http://schemas.google.com/g/2005' 
       xmlns:gCal='http://schemas.google.com/gCal/2005'>
  <title type='text'></title>
  <summary type='text'></summary>
  <gCal:timezone value=''></gCal:timezone>
  <gCal:hidden value=''></gCal:hidden>
  <gCal:color value=''></gCal:color>
  <gd:where rel='' label='' valueString=''></gd:where>
  </entry>"
  
  ACL_XML = "<entry xmlns='http://www.w3.org/2005/Atom' xmlns:gAcl='http://schemas.google.com/acl/2007'>
                      <category scheme='http://schemas.google.com/g/2005#kind' term='http://schemas.google.com/acl/2007#accessRule'/>
                       <gAcl:scope type='default'></gAcl:scope>
                       <gAcl:role value=''></gAcl:role>
                    </entry>"
                    
  EVENT_XML = "<entry xmlns='http://www.w3.org/2005/Atom' xmlns:gd='http://schemas.google.com/g/2005'>
  <category scheme='http://schemas.google.com/g/2005#kind' term='http://schemas.google.com/g/2005#event'></category>
  <title type='text'></title>
  <content type='text'></content>
  <gd:transparency value=''></gd:transparency>
  <gd:eventStatus value=''></gd:eventStatus>
  <gd:where valueString=''></gd:where>
  <gd:when startTime='' endTime=''></gd:when>
</entry>
  "
  
  class AuthenticationFailed < StandardError; end #:nodoc: all

  class NotAuthenticated < StandardError; end
    
  class InvalidService < StandardError; end
    
  class HTTPPostFailed < StandardError; end
    
  class HTTPPutFailed < StandardError; end
    
  class HTTPGetFailed < StandardError; end
    
  class HTTPDeleteFailed < StandardError; end
    
  class CalendarSaveFailed < StandardError; end
  
  class EventSaveFailed < StandardError; end
    
  class RecurrenceValueError < StandardError; end
    
  class CalendarNotEditable < StandardError; end
    
  class QueryParameterError < StandardError; end

  #The ProxyInfo class contains information for configuring a proxy connection

  class ProxyInfo
    attr_accessor :address, :port, :username, :password
    @address = nil
    @port = nil
    @username = nil
    @password = nil

    #The initialize function accepts four variables for configuring the ProxyInfo object.  
    #The proxy connection is initiated using the builtin Net::HTTP proxy support.

    def initialize(address, port, username=nil, password=nil)
      @address = address
      @port = port
      @username = username
      @password = password
    end
  end
  
  #The Base class includes the basic HTTP methods for sending and receiving 
  #messages from the Google Calendar API.  You shouldn't have to use this class 
  #directly, rather access the functionality through the Service subclass.

  class Base
    AUTH_URL = "https://www.google.com/accounts/ClientLogin"
    CALENDAR_LIST_FEED = "http://www.google.com/calendar/feeds/default/allcalendars/full"
    @proxy_info = nil
    @auth_token = nil
    @debug = false

    #Contains the ProxyInfo object for using a proxy server
    attr_accessor :proxy_info
    
    #If set to true, debug will dump all raw HTTP requests and responses
    attr_accessor :debug
    
    # Sends an HTTP POST request.  The header should be a hash of name/value pairs.  
    # Returns the Net::HTTPResponse object on succces, or raises the appropriate
    # error if a non 20x response code is received.
    def send_post(url, content, header=nil)
      header = auth_header(header)
      ret = nil
      location = URI.parse(url)
      puts "url = "+url if @debug
      ret = do_post(location, header, content)
      while ret.is_a?(Net::HTTPRedirection)
        puts "Redirect received, resending post" if @debug
        if oauth?
          header ||= {}
          header['Cookie'] = ret['set-cookie'] if ret['set-cookie']
        end
        ret = do_post(ret['location'], header, content)
      end
      if ret.is_a?(Net::HTTPSuccess)
        puts "20x response received\nResponse: \n"+ret.read_body if @debug
        return ret
      else
        puts "invalid response received: "+ret.code if @debug
        raise HTTPPostFailed, ret.body
      end
    end
    
    def do_post(url, header, content)
      ret = nil
      if url.is_a?(String)
        location = URI.parse(url)
      else
        location = url
      end
      
      puts "Starting post\nHeader: #{header}\n" if @debug
      if oauth?
        header ||= {}
        header.merge!({ 'Accept'=>'application/atom+xml', 'Content-Type' => 'application/atom+xml' })
        ret = @auth_token.post(url.to_s, content, header)
      else
        http = get_http_object(location)
        http.start do |ht|
          ret = ht.post(location.to_s, content, header)
        end
      end
      return ret
    end
    
    # Sends an HTTP PUT request.  The header should be a hash of name/value pairs.  
    # Returns the Net::HTTPResponse object on succces, or raises the appropriate
    # error if a non 20x response code is received.
    def send_put(url, content, header=nil)
      header = auth_header(header)
      ret = nil
      location = URI.parse(url)
      puts "url = "+url if @debug
      ret = do_put(location, header, content)
      while ret.is_a?(Net::HTTPRedirection)
        puts "Redirect received, resending post" if @debug
        ret = do_put(ret['location'], header, content)
      end
      if ret.is_a?(Net::HTTPSuccess)
        puts "20x response received\nResponse: \n"+ret.read_body if @debug
        return ret
      else
        puts "invalid response received: "+ret.code if @debug
        raise HTTPPutFailed, ret.body
      end
    end
    
    def do_put(url, header, content)
      ret = nil
      if url.is_a?(String)
        location = URI.parse(url)
      else
        location = url
      end
      
      puts "Starting put\nHeader: #{header}\n" if @debug
      if oauth?
        header ||= {}
        header.merge!({ 'Accept'=>'application/atom+xml', 'Content-Type' => 'application/atom+xml' })
        ret = @auth_token.put(url.to_s, content, header)
      else
        http = get_http_object(location)
        http.start do |ht|
          ret = ht.put(location.to_s, content, header)
        end
      end
      return ret
    end

    # Sends an HTTP GET request.  The header should be a hash of name/value pairs.  
    # Returns the Net::HTTPResponse object on succces, or raises the appropriate
    # error if a non 20x response code is received.
    def send_get(url, header = nil)
      header = auth_header(header)
      ret = nil
      location = URI.parse(url)
      puts "url = "+url if @debug
      ret = do_get(location, header)

      while ret.is_a?(Net::HTTPRedirection)
        puts "Redirect received from #{location.to_s}, resending get to #{ret['location']}" if @debug
        if oauth?
          header ||= {}
          header['Cookie'] = ret['set-cookie'] if ret['set-cookie']
        end
        ret = do_get(ret['location'], header)
      end
      if ret.is_a?(Net::HTTPSuccess)
        puts "20x response received\nResponse: \n"+ret.read_body if @debug
        return ret
      else
        puts "Error received, resending get" if @debug
        raise HTTPGetFailed, ret.body
      end
    end
    
    def do_get(url, header)
      ret = nil
      if url.is_a?(String)
        location = URI.parse(url)
      else
        location = url
      end

      puts "Starting get\nHeader: #{header}\n" if @debug
      if oauth?
        ret = @auth_token.get(url.to_s, header)
      else
        http = get_http_object(location)
        http.start do |ht|
          ret = ht.get(location.to_s, header)
        end
      end
      return ret
    end

    # Sends an HTTP DELETE request.  The header should be a hash of name/value pairs.  
    # Returns the Net::HTTPResponse object on succces, or raises the appropriate
    # error if a non 20x response code is received.
    def send_delete(url, header = nil)
      header = auth_header(header)
      ret = nil
      location = URI.parse(url)
      puts "url = "+url if @debug
      ret = do_delete(location, header)
      while ret.is_a?(Net::HTTPRedirection)
        puts "Redirect received, resending post" if @debug
        ret = do_delete(ret['location'], header)
      end
      if ret.is_a?(Net::HTTPSuccess)
        puts "20x response received\nResponse: \n"+ret.read_body if @debug
        return true
      else
        puts "invalid response received: "+ret.code if @debug
        raise HTTPDeleteFailed, ret.body
      end
    end
    
    def do_delete(url, header)
      ret = nil
      if url.is_a?(String)
        location = URI.parse(url)
      else
        location = url
      end
      
      puts "Starting get\nHeader: #{header}\n" if @debug
      if oauth?
        ret = @auth_token.delete(url.to_s, header)
      else
        http = get_http_object(location)
        http.start do |ht|
          ret = ht.delete(location.to_s, header)
        end
      end
      return ret
    end

    private

    def get_http_object(location)
      if @proxy_info and @proxy_info.address
	     http = Net::HTTP.new(location.host, location.port, @proxy_info.address, @proxy_info.port, @proxy_info.username, @proxy_info.password)
      else
	     http = Net::HTTP.new(location.host, location.port)
	    end
      if location.scheme == 'https'
        #fixed http/http misnaming via JohnMetta
        puts "SSL True" if @debug
        http.use_ssl = true
        http.verify_mode = OpenSSL::SSL::VERIFY_NONE
      end
      return http
    end

    def auth_header_client_login(header)
      if @auth_token
        if header
          header.merge!({'Authorization' => "GoogleLogin auth=#{@auth_token}", "GData-Version" => "2.1"})
        else 
          header = {'Authorization' => "GoogleLogin auth=#{@auth_token}", "GData-Version" => "2.1"}
        end
      end
      
      return header
    end
    
    #auth header that uses google authsub 
    def auth_header_authsub(header)
      if @auth_token
      	if header
      	  header.merge!({'Authorization' => "AuthSub token=#{@auth_token}", "GData-Version" => "2.1"})
      	else 
      	  header = {'Authorization' => "AuthSub token=#{@auth_token}", "GData-Version" => "2.1"}
      	end
      end
      return header
    end
    
    # routes the authentication to the proper header based upon auth type
    def auth_header(header)
      if @auth_type == 'AuthSub'
        return auth_header_authsub(header)
      elsif @auth_type == 'ClientLogin'
        return auth_header_client_login(header)
      end
    end
    
  end
end

lib/gcal4ruby/event.rb
require 'gcal4ruby/recurrence'

module GCal4Ruby
  #The Event Class represents a remote event in calendar. 
  #
  #=Usage
  #All usages assume a successfully authenticated Service and valid Calendar.
  #1. Create a new Event
  #    event = Event.new(calendar)
  #    event.title = "Soccer Game"
  #    event.start = Time.parse("12-06-2009 at 12:30 PM")
  #    event.end = Time.parse("12-06-2009 at 1:30 PM")
  #    event.where = "Merry Playfields"
  #    event.save
  #
  #2. Find an existing Event
  #    event = Event.find(cal, "Soccer Game", {:scope => :first})
  #
  #3. Find all events containing the search term
  #    event = Event.find(cal, "Soccer Game")
  #
  #4. Create a recurring event for every saturday
  #    event = Event.new(calendar)
  #    event.title = "Baseball Game"
  #    event.where = "Municipal Stadium"
  #    event.recurrence = Recurrence.new
  #    event.recurrence.start = Time.parse("13-06-2009 at 4:30 PM")
  #    event.recurrence.end = Time.parse("13-06-2009 at 6:30 PM")
  #    event.recurrence.frequency = {"weekly" => ["SA"]}
  #    event.save 
  #
  #5. Create an event with a 15 minute email reminder
  #    event = Event.new(calendar)
  #    event.title = "Dinner with Kate"
  #    event.start = Time.parse("20-06-2009 at 5 pm")
  #    event.end = Time.parse("20-06-209 at 8 pm")
  #    event.where = "Luigi's"
  #    event.reminder = {:minutes => 15, :method => 'email'}
  #    event.save
  #
  #6. Create an event with attendees
  #    event = Event.new(calendar)
  #    event.title = "Dinner with Kate"
  #    event.start = Time.parse("20-06-2009 at 5 pm")
  #    event.end = Time.parse("20-06-209 at 8 pm")
  #    event.attendees => {:name => "Kate", :email => "kate@gmail.com"}
  #    event.save
  #
  #After an event object has been created or loaded, you can change any of the 
  #attributes like you would any other object.  Be sure to save the event to write changes
  #to the Google Calendar service.
  class Event
    #The event title
    attr_accessor :title
    #The content for the event
    attr_accessor :content
    #The location of the event
    attr_accessor :where
    #A flag for whether the event show as :free or :busy
    attr_accessor :transparency
    #A flag indicating the status of the event.  Values can be :confirmed, :tentative or :cancelled
    attr_accessor :status
    #The unique event ID
    attr_accessor :id
    #Flag indicating whether it is an all day event
    attr_accessor :all_day
    
    @attendees
    
    #The event start time
    attr_reader :start
    #The event end time
    attr_reader :end
    #The reminder settings for the event, returned as a hash
    attr_reader :reminder
    #The date the event was created
    attr_reader :published
    #The date the event was last updated
    attr_reader :updated
    #The date the event was last edited
    attr_reader :edited
    
    #Sets the reminder options for the event.  Parameter must be a hash containing one of 
    #:hours, :minutes and :days, which are simply the number of each before the event start date you'd like to 
    #receive the reminder.  
    #
    #:method can be one of the following:
    #- <b>'alert'</b>: causes an alert to appear when a user is viewing the calendar in a browser
    #- <b>'email'</b>: sends the user an email message
    def reminder=(r)
      @reminder = r
    end
    
    #Returns the current event's Recurrence information
    def recurrence
      @recurrence
    end
    
    #Returns an array of the current attendees
    def attendees
      @attendees
    end
    
    #Accepts an array of email address/name pairs for attendees.  
    #  [{:name => 'Mike Reich', :email => 'mike@seabourneconsulting.com'}]
    #The email address is requried, but the name is optional
    def attendees=(a)
      if a.is_a?(Array)
        @attendees = a
      else
        raise "Attendees must be an Array of email/name hash pairs"
      end
    end
    
    #Sets the event's recurrence information to a Recurrence object.  Returns the recurrence if successful,
    #false otherwise
    def recurrence=(r)
      if r.is_a?(Recurrence) or r.nil?
        r.event = self unless r.nil?
        @recurrence = r
      else
        return false
      end
    end
    
    #Returns a duplicate of the current event as a new Event object
    def copy()
      e = Event.new()
      e.load(to_xml)
      e.calendar = @calendar
      return e
    end
    
    #Sets the start time of the Event.  Must be a Time object or a parsable string representation
    #of a time.
    def start=(str)
      if str.is_a?String
        @start = Time.parse(str)      
      elsif str.is_a?Time
        @start = str
      else
        raise "Start Time must be either Time or String"
      end
    end
    
    #Sets the end time of the Event.  Must be a Time object or a parsable string representation
    #of a time.
    def end=(str)
      if str.is_a?String
        @end = Time.parse(str)      
      elsif str.is_a?Time
        @end = str
      else
        raise "End Time must be either Time or String"
      end
    end
    
    #Deletes the event from the Google Calendar Service.  All values are cleared.
    def delete
        if @exists    
          if @calendar.service.send_delete(@edit_feed, {"If-Match" => @etag})
            @exists = false
            @deleted = true
            @title = nil
            @content = nil
            @id = nil
            @start = nil
            @end = nil
            @transparency = nil
            @status = nil
            @where = nil
            return true
          else
            return false
          end
        else
          return false
        end
    end
    
    #Creates a new Event.  Accepts a valid Calendar object and optional attributes hash.
    def initialize(calendar, attributes = {})
      if not calendar.editable
        raise CalendarNotEditable
      end
      super()
      attributes.each do |key, value|
        self.send("#{key}=", value)
      end
      @xml ||= EVENT_XML
      @calendar ||= calendar
      @transparency ||= "http://schemas.google.com/g/2005#event.opaque"
      @status ||= "http://schemas.google.com/g/2005#event.confirmed"
      @attendees ||= []
      @all_day ||= false
    end
    
    #If the event does not exist on the Google Calendar service, save creates it.  Otherwise
    #updates the existing event data.  Returns true on success, false otherwise.
    def save
      if @deleted
        return false
      end
      if @exists 
        ret = @calendar.service.send_put(@edit_feed, to_xml, {'Content-Type' => 'application/atom+xml', "If-Match" => @etag})
      else
        ret = @calendar.service.send_post(@calendar.event_feed, to_xml, {'Content-Type' => 'application/atom+xml'})
      end
      if !@exists
        if load(ret.read_body)
          return true
        else
          raise EventSaveFailed
        end
      end
      reload
      return true
    end
    
    #Returns an XML representation of the event.
    def to_xml()
      xml = REXML::Document.new(@xml)
      xml.root.elements.each(){}.map do |ele|
        case ele.name
        when 'id'
          ele.text = @id
        when "title"
          ele.text = @title
        when "content"
          ele.text = @content
        when "when"
          if not @recurrence
            ele.attributes["startTime"] = @all_day ? @start.strftime("%Y-%m-%d") : @start.xmlschema
            ele.attributes["endTime"] = @all_day ? @end.strftime("%Y-%m-%d") : @end.xmlschema
            set_reminder(ele)
          else
            if not @reminder
              xml.root.delete_element("/entry/gd:when")
              xml.root.add_element("gd:recurrence").text = @recurrence.to_s
            else
              ele.delete_attribute('startTime')
              ele.delete_attribute('endTime')
              set_reminder(ele)  
            end
          end
        when "eventStatus"
          ele.attributes["value"] = case @status
            when :confirmed
              "http://schemas.google.com/g/2005#event.confirmed"
            when :tentative
              "http://schemas.google.com/g/2005#event.tentative"
            when :cancelled
              "http://schemas.google.com/g/2005#event.canceled"
            else
              "http://schemas.google.com/g/2005#event.confirmed"
          end
        when "transparency"
          ele.attributes["value"] = case @transparency
              when :free
                "http://schemas.google.com/g/2005#event.transparent"
              when :busy
                "http://schemas.google.com/g/2005#event.opaque"
              else
                "http://schemas.google.com/g/2005#event.opaque"
            end
        when "where"
          ele.attributes["valueString"] = @where
        when "recurrence"
          puts 'recurrence element found' if @calendar.service.debug
          if @recurrence
            puts 'setting recurrence' if @calendar.service.debug
            ele.text = @recurrence.to_s
          else
            puts 'no recurrence, adding when' if @calendar.service.debug
            w = xml.root.add_element("gd:when")
            xml.root.delete_element("/entry/gd:recurrence")
            w.attributes["startTime"] = @all_day ? @start.strftime("%Y-%m-%d") : @start.xmlschema
            w.attributes["endTime"] = @all_day ? @end.strftime("%Y-%m-%d") : @end.xmlschema
            set_reminder(w)
          end
        end
      end        
      if not @attendees.empty?
        @attendees.each do |a|
          xml.root.add_element("gd:who", {"email" => a[:email], "valueString" => a[:name], "rel" => "http://schemas.google.com/g/2005#event.attendee"})
        end
      end
      xml.to_s
    end
    
    #Loads the event info from an XML string.
    def load(string)
      @xml = string
      @exists = true
      xml = REXML::Document.new(string)
      @etag = xml.root.attributes['etag']
      xml.root.elements.each(){}.map do |ele|
          case ele.name
             when 'updated'
                @updated = ele.text
             when 'published'
                @published = ele.text
             when 'edited'
                @edited = ele.text
             when 'id'
                @id, @edit_feed = ele.text
             when 'title'
                @title = ele.text
              when 'content'
                @content = ele.text
              when "when"
                @start = Time.parse(ele.attributes['startTime'])
                @end = Time.parse(ele.attributes['endTime'])
                ele.elements.each("gd:reminder") do |r|
                  @reminder = {:minutes => r.attributes['minutes'] ? r.attributes['minutes'] : 0, :hours => r.attributes['hours'] ? r.attributes['hours'] : 0, :days => r.attributes['days'] ? r.attributes['days'] : 0, :method => r.attributes['method'] ? r.attributes['method'] : ''}
                end
              when "where"
                @where = ele.attributes['valueString']
              when "link"
                if ele.attributes['rel'] == 'edit'
                  @edit_feed = ele.attributes['href']
                end
              when "who"
                if ele.attributes['rel'] == "http://schemas.google.com/g/2005#event.attendee"
                n = {}
                ele.attributes.each do |name, value|
                    case name
                      when "email"
                        n[:email] = value
                      when "valueString"
                        n[:name] = value
                    end
                  end                
               @attendees << n
               end
              when "eventStatus"
              case ele.attributes["value"] 
                when "http://schemas.google.com/g/2005#event.confirmed"
                 @status =  :confirmed
                when "http://schemas.google.com/g/2005#event.tentative"
                  @status = :tentative
                when "http://schemas.google.com/g/2005#event.cancelled"
                  @status = :cancelled
              end
            when 'recurrence'
              @recurrence = ele.text #Recurrence.new(ele.text)
            when "transparency"
               case ele.attributes["value"]
                  when "http://schemas.google.com/g/2005#event.transparent" 
                    @transparency = :free
                  when "http://schemas.google.com/g/2005#event.opaque"
                    @transparency = :busy
                end
            end      
        end
    end
    
    #Reloads the event data from the Google Calendar Service.  Returns true if successful,
    #false otherwise.
    def reload
      t = Event.find(@calendar, @id)
      if t
        if load(t.to_xml)
         return true
        else
         return false
        end
      else
        return false
      end
    end
    
    #Finds the event that matches a query term in the event title or description.
    #  
    #'query' is a string to perform the search on or an event id.
    # 
    #The params hash can contain the following hash values
    #* *scope*: may be :all or :first, indicating whether to return the first record found or an array of all records that match the query.  Default is :all.
    #* *range*: a hash including a :start and :end time to constrain the search by
    #* *max_results*: an integer indicating the number of results to return.  Default is 25.
    #* *sort_order*: either 'ascending' or 'descending'.
    #* *single_events*: either 'true' to return all recurring events as a single entry, or 'false' to return all recurring events as a unique event for each recurrence.
    #* *ctz*: the timezone to return the event times in
    def self.find(calendar, query = '', params = {})
      query_string = ''
      
      begin 
        test = URI.parse(query).scheme
      rescue Exception => e
        test = nil
      end
      
      if test
        puts "id passed, finding event by id" if calendar.service.debug
        puts "id = "+query if calendar.service.debug
        event_id = query.gsub("/events/","/private/full/") #fix provided by groesser3
      
        es = calendar.service.send_get(event_id)
        puts es.inspect if calendar.service.debug
        if es
          entry = REXML::Document.new(es.read_body).root
          puts 'event found' if calendar.service.debug
          Event.define_xml_namespaces(entry)
          event = Event.new(calendar)
          event.load("<?xml version='1.0' encoding='UTF-8'?>#{entry.to_s}")
          return event
        end
        return nil
      end

  
      #parse params hash for values
      range = params[:range] || nil
      max_results = params[:max_results] || nil
      sort_order = params[:sortorder] || nil
      single_events = params[:singleevents] || nil
      timezone = params[:ctz] || nil
      
      #set up query string
      query_string += "q=#{CGI.escape(query)}" if query
      if range
        if not range.is_a? Hash or (range.size > 0 and (not range[:start].is_a? Time or not range[:end].is_a? Time))
          raise "The date range must be a hash including the :start and :end date values as Times"
        else
          date_range = ''
          if range.size > 0
            #Added via patch from Fabio Inguaggiato
            query_string += "&start-min=#{CGI::escape(range[:start].xmlschema)}&start-max=#{CGI::escape(range[:end].xmlschema)}"
          end
        end
      end
      query_string += "&max-results=#{max_results}" if max_results
      query_string += "&sortorder=#{sort_order}" if sort_order
      query_string += "&ctz=#{timezone.gsub(" ", "_")}" if timezone
      query_string += "&singleevents=#{single_events}" if single_events
      if query_string
        events = calendar.service.send_get("http://www.google.com/calendar/feeds/#{calendar.id}/private/full?"+query_string)
        ret = []
        REXML::Document.new(events.read_body).root.elements.each("entry"){}.map do |entry|
          Event.define_xml_namespaces(entry)
          event = Event.new(calendar)
          event.load("<?xml version='1.0' encoding='UTF-8'?>#{entry.to_s}")
          ret << event
        end
      end
      if params[:scope] == :first
        return ret[0]
      else
        return ret
      end
    end
    
    #Returns true if the event exists on the Google Calendar Service.
    def exists?
      return @exists
    end
  
    private 
    @exists = false
    @calendar = nil
    @xml = nil
    @etag = nil
    @recurrence = nil
    @deleted = false
    @edit_feed = ''
    
    def self.define_xml_namespaces(entry)
      entry.attributes["xmlns:gCal"] = "http://schemas.google.com/gCal/2005"
      entry.attributes["xmlns:gd"] = "http://schemas.google.com/g/2005"
      entry.attributes["xmlns:app"] = "http://www.w3.org/2007/app"
      entry.attributes["xmlns"] = "http://www.w3.org/2005/Atom"
      entry.attributes["xmlns:georss"] = "http://www.georss.org/georss"
      entry.attributes["xmlns:gml"] = "http://www.opengis.net/gml"
    end
    
    def set_reminder(ele)
      ele.delete_element("gd:reminder")
      if @reminder
        e = ele.add_element("gd:reminder")
        used = false
        if @reminder[:minutes] 
          e.attributes['minutes'] = @reminder[:minutes] 
          used = true
        elsif @reminder[:hours] and not used
          e.attributes['hours'] = @reminder[:hours]
          used = true
        elsif @reminder[:days] and not used
          e.attributes['days'] = @reminder[:days]
        end
        if @reminder[:method] 
          e.attributes['method'] = @reminder[:method]
        else
          e.attributes['method'] = 'email'
        end
      else
        ele.delete_element("gd:reminder")
      end
    end
  end
end

